src/lib/io/sol-uart-impl-linux.c

```diff
@@ -173,71 +321,60 @@ SOL_API void
 sol_uart_close(struct sol_uart *uart)
 {
     SOL_NULL_CHECK(uart);
-    if (uart->async.rx_fd_handler)
-        sol_fd_del(uart->async.rx_fd_handler);
-    if (uart->async.tx_fd_handler)
-        sol_fd_del(uart->async.tx_fd_handler);
-    close(uart->fd);
-    free(uart);
-}
+    SOL_EXP_CHECK(uart->delete_me);
 
-static void
-uart_tx_dispatch(struct sol_uart *uart, int status)
-{
-    uart->async.tx_fd_handler = NULL;
-    if (!uart->async.tx_cb)
+    if (uart->in_use) {
```

Comment:
before this line you should delete the sol_fd/sol_timeout, not inside the inner `close_uart()`